metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/FunctionResolverSingleton.java

```diff
@@ -0,0 +1,206 @@
+  private static final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private static FunctionResolverSingleton INSTANCE = new FunctionResolverSingleton();
+
+  private FunctionResolverSingleton() {}
+
+  public static FunctionResolver getInstance() {
+    return INSTANCE;
+  }
+
+
+
+  @Override
+  public Iterable<StellarFunctionInfo> getFunctionInfo() {
+    return _getFunctions().values();
+  }
+
+  @Override
+  public Iterable<String> getFunctions() {
+    return _getFunctions().keySet();
+  }
+
+  @Override
+  public void initialize(Context context) {
+    //forces a load of the stellar functions.
+    _getFunctions();
+  }
+
+  /**
+   * This allows the lazy loading of the functions.  We do not want to take a multi-second hit to analyze the full classpath
+   * every time a unit test starts up.  That would cause the runtime of things to blow right up.  Instead, we only want
+   * to take the hit if a function is actually called from a stellar expression.
+   *
+   *
+   * @return The map of the stellar functions that we found on the classpath indexed by fully qualified name
+   */
+  private Map<String, StellarFunctionInfo> _getFunctions() {
+    /*
+     * Because we are not doing this in a static block and because this object is a singleton we have to concern ourselves with
+     * the possiblity that two threads are calling this function at the same time.  Normally, I would consider just making the
+     * function synchronized, but since every stellar statement which uses a function will be here, I wanted to distinguish
+     * between read locks (that happen often and are quickly resolved) and write locks (which should happen at initialization time).
+     */
+    lock.readLock().lock();
+    try {
+      if (isInitialized.get()) {
+        return functions;
+      }
+    }
+    finally {
+      lock.readLock().unlock();
+    }
+    //we should VERY rarely get here.
+    lock.writeLock().lock();
+    try {
+      //I have to check again because two threads or more could be waiting at the lock statement.  The loser threads
+      //shouldn't reinitialize.
+      if(!isInitialized.get()) {
+        loadFunctions(functions);
+        isInitialized.set(true);
+      }
+      return functions;
+    }
+    finally {
+      lock.writeLock().unlock();
+    }
+  }
+  /**
+   * Applies this function to the given argument.
+   *
+   * @param s the function argument
+   * @return the function result
+   */
+  @Override
+  public StellarFunction apply(String s) {
+    StellarFunctionInfo ret = _getFunctions().get(s);
+    if(ret == null) {
+      throw new IllegalStateException("Unable to resolve function " + s);
+    }
+    return ret.getFunction();
+  }
+
+  private void loadFunctions(final Map<String, StellarFunctionInfo> ret) {
+    try {
+      ClassLoader classLoader = getClass().getClassLoader();
+      Reflections reflections = new Reflections(new ConfigurationBuilder().setUrls(effectiveClassPathUrls(classLoader)));
```

Comment:
Yes, I think it would be a good follow-on.  I really just bring it up as a discussion point to think about.  Good idea, bad idea?